cmd/localapiclient/localapiclient.go

```diff
@@ -0,0 +1,105 @@
+package localapiclient
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"tailscale.com/ipn/localapi"
+)
+
+// Response represents the result of processing an http.Request.
+type Response struct {
+	headers              http.Header
+	status               int
+	bodyWriter           net.Conn
+	bodyReader           net.Conn
+	startWritingBody     chan interface{}
+	startWritingBodyOnce sync.Once
+}
+
+func (r *Response) Header() http.Header {
+	return r.headers
+}
+
+// Write writes the data to the response body and will send the data to Java.
+func (r *Response) Write(data []byte) (int, error) {
+	r.Flush()
+	if r.status == 0 {
+		r.WriteHeader(http.StatusOK)
+	}
+	return r.bodyWriter.Write(data)
+}
+
+func (r *Response) WriteHeader(statusCode int) {
+	r.status = statusCode
+}
+
+func (r *Response) Body() net.Conn {
+	return r.bodyReader
+}
+
+func (r *Response) StatusCode() int {
+	return r.status
+}
+
+func (r *Response) Flush() {
+	r.startWritingBodyOnce.Do(func() {
+		close(r.startWritingBody)
+	})
+}
+
+type LocalAPIClient struct {
+	h *localapi.Handler
+}
+
+func New(h *localapi.Handler) *LocalAPIClient {
+	return &LocalAPIClient{h: h}
+}
+
+// Call calls the given endpoint on the local API using the given HTTP method
+// optionally sending the given body. It returns a Response representing the
+// result of the call and an error if the call could not be completed or the
+// local API returned a status code in the 400 series or greater.
+// Note - Response includes a response body available from the Body method, it
+// is the caller's responsibility to close this.
+func (cl *LocalAPIClient) Call(ctx context.Context, deadline time.Time, method, endpoint string, body io.Reader) (*Response, error) {
```

Comment:
The idiomatic way to handle this is to have the caller set a deadline on the context instead of passing it into here. If you need to get the context's deadline, you can do this

```
deadline, hasDeadline := ctx.Deadline()
```